dice_ml/explainer_interfaces/dice_KD.py

```diff
@@ -96,7 +96,7 @@ def _generate_counterfactuals(self, query_instance, total_CFs, desired_range=Non
         self.dataset_with_predictions, self.KD_tree, self.predictions = self.build_KD_tree(data_df_copy, desired_range,
                                                                                            desired_class, self.predicted_outcome_name)
 
-        query_instance, final_cfs, cfs_preds = self.find_counterfactuals(data_df_copy,
+        query_instance, cfs_preds = self.find_counterfactuals(data_df_copy,
```

Comment:
The `find_counterfactuals` needs to be refactored. This is a larger problem with other files also--perhaps we can take this up in a future PR. Ideally it should return the final_cfs and then we should assign it to self.final_cfs here in this function. The find_counterfactuals shouldn't be setting `self` attributes.

Reply:
Yes, I agree completely! Functions ideally shouldn't change any self variables and should return everything that is changed